### ISSUE TITLE
BetterEnd not loading the edited level.dat

### ISSUE THREAD

**Issue by Alkiasan** [paulevsGitch/BetterEnd#208]
Hi, I'm having an issue with loading the level.dat that I have edited following your tutorial.
I tried following the steps exactly, but for some reason, my server throws an error reading the level.dat on launch.
Interestingly, the world still loads, but all the gamerules,worldborder, etc. reset.
[level.zip](https://github.com/paulevsGitch/BetterEnd/files/6682254/level.zip)
Above are both dat files, the edited one (level.dat) and the backup (backup.dat)
Here is the error that prints on server startup:
https://gist.github.com/Alkiasan/6b5b3ab2fe206032ba2b382e78710861
Any help with this would be appreciated, no clue why the dat file isn't working..

**Comment by Alkiasan** [comment]
Should probably note that we reset the end dimension for this update by deleting the DIM1 folder, and are loading this on a server that didn't have the BetterEnd mod before, if that's got anything to do with it.
I can send the entire server file if you want, but it is 14GB so it may take a while..

**Comment by paulevsGitch** [comment]
BetterEnd is not loading level.dat, it is loaded by Minecraft itself. If it is not loaded correctly - this means editing mistake, MC bug, datapack conflict, etc.

> my server throws an error reading the level.dat on launch.

From your logs I can assume that it is wrong biome codec for one of your biomes. Do you have other mods that adds biomes into the End or change it's generation? This can be a potential mod conflict

**Comment by Alkiasan** [comment]
I have no other mods that affect end generation.
Only datapacks that might do SOMETHING are the datapacks that add custom dimensions.
Here's my mods list:

Adorn
Blockus
Chunky
Croptopia
Deltalogger
DisguiseLib
DualRiders
EmoteCraft
Origins
Carpet 
Fabric-discord-link
Fatxporbs
FabricTailor
Flan
Haema
Invisiblights
Invview
Krypton
LazyDFU
Lithium
NotEnoughAnimations
Pehkui
Player Roles
Taterzens
SimpleVoiceChat
Vanish (DOESNT WORK ON RELEASE 1.17)
Extra Origins
Origins Classes
Slime Origin
SimpleHarvest
Spark
AntiqueAtlas
Better Graves
Cosmetic Armor
Curios API
Transparent Armor
MobDisguises
Drogtor the nickinator
FantasyCraft
SimpleHarvest
HorizonRP (Custom mod, just adds items)
Image2Map
FastLeafDecay
Origins Umbrellas
RandomPatches
Text Colorizer plugin
Xaero's worldmap (Coming soon, Minimap was JUST Updated.)
Worldedit

**Comment by Alkiasan** [comment]
Oh! Fantasycraft adds specific ores to the end, could that be causing it?

**Comment by paulevsGitch** [comment]
Probably, can you please test on a separate world disabling different mods and datapacks? Actually, this is the only one way to find conflict if they exists

**Comment by Alkiasan** [comment]
The weird part is that level.dat loads fine on separate worlds with the same mods..

**Comment by paulevsGitch** [comment]
Looks like the reason is in datapacks - they can cause such issues

**Comment by Alkiasan** [comment]
Luckily, I just created a backup of the world the server is using so I'll try loading it without fantasycraft and see if the issue persists

**Comment by Alkiasan** [comment]
Tried without fantasycraft. Issue persists.
Interestingly, loading in singleplayer causes this:
![image](https://user-images.githubusercontent.com/62717881/122679517-17ae1c80-d1e3-11eb-81df-560dfc7cb018.png)
But we have no datapacks that effect end generation...
Not only that, but the issue only started when betterend was installed. Beforehand, the datapacks worked fine.
Here's a list of datapacks if it helps
![image](https://user-images.githubusercontent.com/62717881/122679549-37454500-d1e3-11eb-9b03-fcf84a4f44d7.png)

**Comment by Alkiasan** [comment]
Bee, Bunborn, DwarfEdit, ExplorerDP, Fairy, Blazeborn, Floran, Meatfix, Vampire_Origin, orboforigin, puppeteer, skeletal are all custom origin datapacks for the origins mod, so I doubt they're causing it.
BlockusBlock, Croptopiasmokerfix, RemoveTable are all recipe datapacks that add/remove recipes, So I doubt they're causing it either...
There are ChimeraDimension, SallanCore and TheVoid that all add custom void dimensions, but they shouldn't touch the end as they create their own seperate dimensions.

**Comment by Alkiasan** [comment]
Just tried removing all datapacks, minecraft still shows that screen saying there was a problem with datapacks, even when none are loaded. Can confirm that it's not datapacks causing it.

**Comment by Alkiasan** [comment]
[latest.log](https://github.com/paulevsGitch/BetterEnd/files/6682611/latest.log)
In other news, here's a log for trying to load in singleplayer, in case there's something different.

**Comment by Alkiasan** [comment]
I have removed every mod apart from betterend and the issue persists.

**Comment by Alkiasan** [comment]
[latest.log](https://github.com/paulevsGitch/BetterEnd/files/6682626/latest.log)
Here is a log using exclusively BetterEnd and fabric API, with the same issue.
Assuming it's something wrong in level.dat then?

**Comment by Alkiasan** [comment]
For a bandaid fix, would pregenning the end dimension in a seperate world then dragging its DIM1 folder into the server work?
I can use a default level.dat, but as soon as i change the generator type it breaks

**Comment by paulevsGitch** [comment]
If it is changing each time on world loading and cause issues I still think it is a data pack issue. It can be hidden functionality of another mod or one of datapacks (as most datapack-related code in MC is a bit weird)

Yes, world pregeneration will work, you only need to copy registry cache from server to prevent biome ID shift (Fabric Dynamic registry in world data folder)

**Comment by Alkiasan** [comment]
But I tried with exclusively betterend and the issue still persisted... it really does just seem to break as soon as the type is changed.

**Comment by paulevsGitch** [comment]
I think this is some kind of corruption as you are using level.dat after all mods and datapacks were added/removed/modified

**Comment by Alkiasan** [comment]
Alright... I'm currently trying the pregen idea Which will probably take a while so I'll get back to you on that.
If I had to guess as to what the incompatibility IS, I'd say its likely with custom dimension datapacks, as I can't see the other dimensions effecting it...
Problem being its hard to test this because removing a custom dimension from a server causes it to freak out, but that's my bet anyways

**Comment by Alkiasan** [comment]
Huh. Pregenning world was working fine pregenning with the custom biome sources, but after a while it just stopped pregenning.
Then, upon re-opening the world the same level.dat issue happens again.
[message (5).txt](https://github.com/paulevsGitch/BetterEnd/files/6688220/message.5.txt)

This error seems to appear, too.

**Comment by paulevsGitch** [comment]
Looks like something corrupted light storage in chunks:
![image](https://user-images.githubusercontent.com/17707380/122800063-2a9c1c00-d2cb-11eb-8ee0-34f48bf4500e.png)
![image](https://user-images.githubusercontent.com/17707380/122800095-32f45700-d2cb-11eb-8213-214a6c12118e.png)

**Comment by Alkiasan** [comment]
Interesting. The PC generating wasn't using any performance mods, so I have no clue as to what was effecting lighting storage...

**Comment by paulevsGitch** [comment]
You can try to install Phosphor, it will probably cab fix light issue.
Can you please also find a starting point of all this messages? Light error can be a side effect of something different

**Comment by Alkiasan** [comment]
Alright. We are trying again on a fresh end dim with the same seed using exclusively betterend and chunky. Seems to be working nicer. I can also send the full log from our previous attempt too with the light errors if you want it?

**Comment by Alkiasan** [comment]
Should probably mention that we're using BetterEnd 0.9.6-pre if that has anything to do with all of this.

**Comment by paulevsGitch** [comment]
Well, each version have at least fixes, but there was no any bug with lighting

**Comment by Alkiasan** [comment]
Alright, and updating on just trying to load a server with the level.dat, in singleplayer and multiplayer, loading the world once with a default level.dat works. However, upon loading a second time, the level.dat no longer gets read.
Minecraft region fixer says it's readable, so I have no clue.
Trying the solution minecraft recommends of "Loading with vanilla pack only" also doesn't work.

**Comment by paulevsGitch** [comment]
You can try create a new world with same seed and installed BE and then copy its level dat into your server world. It will replace (probably) corrupted level.dat with working one. I can also recommend try to disable all datapacks before this operation and then add them one by one after it and watch if issue will appear again

**Comment by Alkiasan** [comment]
Okay, so far, it loads without any datapacks, however, player positions/inventories are reset, is that normal?

**Comment by paulevsGitch** [comment]
For singleplayer - yes (singleplayer inventory stored in level.dat), for multiplayer - no

**Comment by Alkiasan** [comment]
I HAVE FOUND THE INCOMPATIBILITY!
After attempting to add a datapack that adds a custom dimension, level.dat is no longer readable.
Now I know that this is a big ask, but is it possible to get compatibility for this?
I can send an example of a dim datapack we was using

**Comment by paulevsGitch** [comment]
I don't know much about datapack dimensions, but I can look at it as I know how datapack biome sources works

**Comment by Alkiasan** [comment]
[ChimeraDimension.zip](https://github.com/paulevsGitch/BetterEnd/files/6688832/ChimeraDimension.zip)
Here's the dim file
Hopefully you can find something to help with this, it'd be a shame if we had to disable custom dimensions...

**Comment by paulevsGitch** [comment]
It is very simple inside - only 2 json files, both are dimension settings, looks more like a demo pack

**Comment by Alkiasan** [comment]
Yes, but for some reason, adding that datapack causes incompatibility issues with BetterEnd

**Comment by paulevsGitch** [comment]
Probably this datapack don't have necessary files to functionate properly or have wrong namespace

**Comment by paulevsGitch** [comment]
You can try [Cardboard](https://curseforge.com/minecraft/mc-mods/bukkit/) which allows to run Spigot plugins on Fabric server and [Multiverse](https://www.spigotmc.org/resources/multiverse-core.390/) plugin, it can be used to create infinity amount of custom dimensions

**Comment by Alkiasan** [comment]
But the datapack works fine without betterend installed. Something with having betterend alongside it isn't working.

**Comment by paulevsGitch** [comment]
It can be a bug in Fabric API itself as Mojang code don't have internal API for mod biome sources and only have datapack support. It can also be a wrong cache files (something similar happens with biomes when two or more biomes get the same ID or 0 ID)

**Comment by Alkiasan** [comment]
So is it unfixable? I don't think we can run cardboard due to incompatibility with our mods.

**Comment by paulevsGitch** [comment]
You can try to find another mod that adds similar dimension or another analogue of Multiverse

**Comment by Alkiasan** [comment]
And will multiverse 100% work alongside betterend? If the issue is just having more than the three main dimensions, wouldn't the issue still persist?

**Comment by paulevsGitch** [comment]
I actually don't know, but I tested BE with Aether which resulted as 4 custom dimensions, and it worked

**Comment by Alkiasan** [comment]
hmmm... The custom dimension itself is a void dimension that has been custom-built, and from what i've seen there's no other alternatives to multiverse, So I guess we'll have to find some sort of way of pasting the custom builds in the void dimension into the overworld..
At least we found the incompatibility, I suppose.

**Comment by paulevsGitch** [comment]
> So I guess we'll have to find some sort of way of pasting the custom builds in the void dimension into the overworld..

[WorldEdit for Fabric](https://www.curseforge.com/minecraft/mc-mods/worldedit)

**Comment by Alkiasan** [comment]
They're big builds is what I mean, it will be quite difficult hahah

**Comment by Alkiasan** [comment]
![image](https://user-images.githubusercontent.com/62717881/122827252-06950680-d2dc-11eb-8cc8-af31902eb270.png)
Would this be causing the issue? BetterEnd adds its structures to the custom dimensions

**Comment by Alkiasan** [comment]
[TheVoid.zip](https://github.com/paulevsGitch/BetterEnd/files/6689552/TheVoid.zip)
For example, this datapack works fine, and the dimension generated by it doesn't have a structures list in level.dat...

**Comment by paulevsGitch** [comment]
Looks like bug in dynamic biome registry, probably some biomes have same ID and this causes issues. You can check registry in world data folder, all biomes in list should have unique IDs

**Comment by Alkiasan** [comment]
![image](https://user-images.githubusercontent.com/62717881/122836506-46fb8100-d2ea-11eb-8a4d-d48a7f75edee.png)
![image](https://user-images.githubusercontent.com/62717881/122836531-51b61600-d2ea-11eb-94b1-696909984748.png)
They all seem to have unique IDs, unless I'm missing something?

**Comment by Alkiasan** [comment]
I was able to fix ChimeraDim.zip (The problematic datapack)
I changed a couple of things;
1. There was an empty structures array inside of dimension settings. I removed that.
2. I changed the biome type from plains to void
3. I changed generator settings to void
3. I defined a custom biome generator and biome sources

I noticed that when the generator settings were set to plains, they'd create a folder that had structures inside of them, but when they was set to void, minecraft would store it as a string instead. Therefore not creating a folder for structures.
This is a bandaid workaround but it works perfectly. level.dat works normally again and everything is back to normal. 
Super weird bug but it is what it is I suppose. At least we found out the problem.
